### PR TITLE
New git-aware cache file layout

### PR DESCRIPTION
**Pull request by julien-c**
## A new way to layout cached files on disk, unifying what we do for `snapshot_download` and `cached_download`, and laying the ground for new features (ability to query the disk to figure out which local files/revisions we have)

See old comment from last year inside `snapshot_download`:

```python
# Note: at some point maybe this format of storage should actually replace
# the flat storage structure we've used so far (initially from allennlp
# if I remember correctly).
```

This new layout will be git-aware (compatible with versioning)

### One picture is worth 1,000 words

Here's a screenshot of the cache's file tree generated by a few file downloads from a model repo (sample code below)

image

```python
import torch
from huggingface_hub.file_download import hf_hub_download

OLDER_REVISION = "bbc77c8132af1cc5cf678da3f1ddf2de43606d48"

hf_hub_download("julien-c/EsperBERTo-small", filename="README.md")

hf_hub_download("julien-c/EsperBERTo-small", filename="pytorch_model.bin")

hf_hub_download(
    "julien-c/EsperBERTo-small", filename="README.md", revision=OLDER_REVISION
)

weights_file = hf_hub_download(
    "julien-c/EsperBERTo-small", filename="pytorch_model.bin", revision=OLDER_REVISION
)

w = torch.load(weights_file, map_location=torch.device("cpu"))
# Yay it works! just loaded a torch file from a symlink
```

### Preliminary mini-spec

- The cache directory contains one subfolder per repo_id (namespaced by repo type)
- inside each repo folder:
  - `refs` is a list of the latest known `revision => commit_hash` pairs
  - `blobs` contains the actual file blobs (identified by their `git-sha` or `sha256`, depending on whether they're LFS files or not)
  - `snapshots` contains one subfolder per commit, each "commit" contains the subset of the files that have been resolved at that particular commit. Each filename is a symlink to the blob at that particular commit.

### Explaining the result

In the sample code I'm downloading the same two filenames at two different commits. 
- The `README.md` is different at both commits so there's both versions of it in `blobs`.
- The `pytorch_model.bin` however, is the same at both commits, so it's not downloaded again, and the disk space is shared between the two snapshots. 🎉 🎉

### CODE REVIEW

File: src/huggingface_hub/snapshot_download.py [hunk 1]
```diff
@@ -178,58 +187,51 @@ def snapshot_download(
...
-        # if we have internet connection we retrieve the correct folder name from the huggingface api
-        _api = HfApi()
-        model_info = _api.model_info(repo_id=repo_id, revision=revision, token=token)
+        return storage_folder
+
+    # if we have internet connection we retrieve the correct folder name from the huggingface api
+    _api = HfApi()
+    repo_info = _api.repo_info(
+        repo_id=repo_id, repo_type=repo_type, revision=revision, token=token
+    )
```
Comment (julien-c) [inline src/huggingface_hub/snapshot_download.py:199]: This uses the new `repo_info()` function from #792

**Comment by julien-c** [comment]
Some additional context for this is in https://github.com/huggingface/transformers/issues/15927#issuecomment-1060664149

**Comment by julien-c** [comment]
## Ok this should be ready for a deeper review @LysandreJik  🎉 

`snapshot_download` is way simpler now and only makes one networking call if you already have the repo snapshot locally 🎉  

Also cc @patrickvonplaten and @sgugger

**Comment by patrickvonplaten** [comment]
Wow this is super cool! No more re-downloading all files when just the README is changed :heart:

File: src/huggingface_hub/snapshot_download.py [hunk 2]
```diff
@@ -120,133 +141,69 @@ def snapshot_download(
...
-    # remove all `/` occurrences to correctly convert repo to directory name
-    repo_id_flattened = repo_id.replace("/", REPO_ID_SEPARATOR)
-
-    # if we have no internet connection we will look for the
-    # last modified folder in the cache
-    if local_files_only:
-        # possible repos have <path/to/cache_dir>/<flatten_repo_id> prefix
-        repo_folders_prefix = os.path.join(cache_dir, repo_id_flattened)
-
-        # list all possible folders that can correspond to the repo_id
```
Comment (patrickvonplaten) [inline src/huggingface_hub/snapshot_download.py:132]: All that beautiful code is gone :cry:

File: src/huggingface_hub/snapshot_download.py [hunk 3]
```diff
@@ -120,133 +141,69 @@ def snapshot_download(
...
-
-        # Create potential nested dir
-        nested_dirname = os.path.dirname(
-            os.path.join(storage_folder, relative_filepath)
+    # if we have no internet connection we will look for an
+    # appropriate folder in the cache
+    # If the specified revision is a commit hash, look inside "snapshots".
+    # If the specified revision is a branch or tag, look inside "refs".
+    if local_files_only:
+        if REGEX_COMMIT_HASH.match(revision):
```
Comment (patrickvonplaten) [inline src/huggingface_hub/snapshot_download.py:158]: (nit) might be easier to understand

```suggestion
        commit_hash = revision
        if not REGEX_COMMIT_HASH.match(commit_hash):
            # rertieve commit_hash from file
            ref_path = os.path.join(storage_folder, "refs", revision)
            with open(ref_path) as f:
                commit_hash = f.read()
        
        snapshot_folder = os.path.join(storage_folder, "snapshots", commit_hash)
        if os.path.exists(snapshot_folder):
            return snapshot_folder
```

            
            
```

File: src/huggingface_hub/snapshot_download.py [hunk 4]
```diff
@@ -120,133 +141,69 @@ def snapshot_download(
...
+        else:
+            ref_path = os.path.join(storage_folder, "refs", revision)
+            with open(ref_path) as f:
+                commit_hash = f.read()
+            snapshot_folder = os.path.join(storage_folder, "snapshots", commit_hash)
+            if os.path.exists(snapshot_folder):
+                return snapshot_folder
+
+        raise ValueError(
+            "Cannot find an appropriate cached folder for the specified revision on the"
```
Comment (patrickvonplaten) [inline src/huggingface_hub/snapshot_download.py:171]:
```suggestion
            "Cannot find an appropriate cached snapshot folder for the specified revision on the"
```

File: src/huggingface_hub/snapshot_download.py [hunk 5]
```diff
@@ -120,133 +141,69 @@ def snapshot_download(
...
-                f" from the branch {revision} in offline mode might lead to unexpected"
-                " behavior by not taking into account the latest commits."
-            )
-
-        # find last modified folder
-        storage_folder = max(repo_folders, key=os.path.getmtime)
+    if repo_type is None:
+        repo_type = "model"
+    if repo_type not in REPO_TYPES:
+        raise ValueError("Invalid repo type")
```
Comment (patrickvonplaten) [inline src/huggingface_hub/snapshot_download.py:147]:
```suggestion
        raise ValueError(f"Invalid repo type. {repo_type} does not exist. Please select one of {REPO_TYPES}")
```

File: src/huggingface_hub/snapshot_download.py [hunk 6]
```diff
@@ -120,133 +141,69 @@ def snapshot_download(
...
-            url,
-            cache_dir=storage_folder,
-            force_filename=relative_filepath,
+    # if we have internet connection we retrieve the correct folder name from the huggingface api
+    _api = HfApi()
+    repo_info = _api.repo_info(
+        repo_id=repo_id, repo_type=repo_type, revision=revision, token=token
+    )
+    filtered_repo_files = _filter_repo_files(
+        repo_files=[f.rfilename for f in repo_info.siblings],
```
Comment (patrickvonplaten) [inline src/huggingface_hub/snapshot_download.py:182]: would this also work with three layers deep?

File: src/huggingface_hub/snapshot_download.py [hunk 7]
```diff
@@ -120,133 +141,69 @@ def snapshot_download(
...
+    # If the specified revision is a branch or tag, look inside "refs".
+    if local_files_only:
+        if REGEX_COMMIT_HASH.match(revision):
+            snapshot_folder = os.path.join(storage_folder, "snapshots", revision)
+            if os.path.exists(snapshot_folder):
+                return snapshot_folder
+        else:
+            ref_path = os.path.join(storage_folder, "refs", revision)
+            with open(ref_path) as f:
+                commit_hash = f.read()
```
Comment (patrickvonplaten) [inline src/huggingface_hub/snapshot_download.py:165]: that's a very nice mechanism!

File: src/huggingface_hub/file_download.py [hunk 8]
```diff
@@ -771,21 +845,260 @@ def hf_hub_download(
...
+            pointer_path = os.path.join(
+                storage_folder, "snapshots", revision, relative_filename
+            )
+            if os.path.exists(pointer_path):
+                return pointer_path
+        else:
+            ref_path = os.path.join(storage_folder, "refs", revision)
+            with open(ref_path) as f:
+                commit_hash = f.read()
+            pointer_path = os.path.join(
```
Comment (patrickvonplaten) [inline src/huggingface_hub/file_download.py:993]: (nit) could maybe simplify the if-else structure here, to just have:

```py
commit_hash = revision
if not REGEX_COMMIT_HASH.match(revision):
            ref_path = os.path.join(storage_folder, "refs", revision)
            with open(ref_path) as f:
                commit_hash = f.read()

pointer_path = os.path.join(
    storage_folder, "snapshots", revision, relative_filename
)
if os.path.exists(pointer_path):
    return pointer_path
```

**Review by patrickvonplaten — APPROVED** [review]
Super excited about this feature. We can advertise it well for all Wav2Vec2 + ngram !

Left only some nits, can all be disregarded

File: src/huggingface_hub/file_download.py [hunk 9]
```diff
@@ -771,21 +845,260 @@ def hf_hub_download(
...
+            etag = _normalize_etag(etag)
+            # In case of a redirect,
+            # save an extra redirect on the request.get call,
+            # and ensure we download the exact atomic version even if it changed
+            # between the HEAD and the GET (unlikely, but hey).
+            if 300 <= r.status_code <= 399:
+                url_to_download = r.headers["Location"]
+        except (requests.exceptions.SSLError, requests.exceptions.ProxyError):
+            # Actually raise for those subclasses of ConnectionError
+            raise
```
Comment (sgugger) [inline src/huggingface_hub/file_download.py:962]: On the Transformers side, we also raise the `RepositoryNotFoundError`, `EntryNotFoundError` and `RevisionNotFoundError` here (which are raised by some code in the try block). There are different from connection/timeout errors and give explicit information to the user about what's wrong, so would be great to integrate those here too.

(Can go in a separate PR if it makes more sense.)

File: src/huggingface_hub/file_download.py [hunk 10]
```diff
@@ -771,21 +845,260 @@ def hf_hub_download(
...
+        logger.info("storing %s in cache at %s", url, blob_path)
+        os.replace(temp_file.name, blob_path)
+
+        logger.info("creating pointer to %s from %s", blob_path, pointer_path)
+        _create_relative_symlink(blob_path, pointer_path)
+
+    try:
+        os.remove(lock_path)
+    except OSError:
+        pass
```
Comment (sgugger) [inline src/huggingface_hub/file_download.py:1102]: Should this be in a `finally` block so it's always executed, even if the previous code failed?

**Review by sgugger — APPROVED** [review]
Super nice new system! Can't wait to be able to use in Transformers :-)

File: src/huggingface_hub/file_download.py [hunk 11]
```diff
@@ -771,21 +845,260 @@ def hf_hub_download(
...
+        with temp_file_manager() as temp_file:
+            logger.info("downloading %s to %s", url, temp_file.name)
+
+            http_get(
+                url_to_download,
+                temp_file,
+                proxies=proxies,
+                resume_size=resume_size,
+                headers=headers,
+            )
```
Comment (LysandreJik) [inline src/huggingface_hub/file_download.py:1091]: In the connection errors I had, this failed and ended up leaving a ~100MB temporary file at the root of the cache; not sure if we can do something about that though

File: src/huggingface_hub/file_download.py [hunk 12]
```diff
@@ -142,6 +147,9 @@ def get_fastcore_version():
     return _fastcore_version
 
 
+REGEX_COMMIT_HASH = re.compile(r"^[0-9a-f]{5,40}$")
```
Comment (LysandreJik) [inline src/huggingface_hub/file_download.py:150]: Theoretically it could be 4 instead of 5 here, as hf.co resolves correctly down to as low as 4 characters
```suggestion
REGEX_COMMIT_HASH = re.compile(r"^[0-9a-f]{4,40}$")
```

File: src/huggingface_hub/file_download.py [hunk 13]
```diff
@@ -689,6 +706,61 @@ def _resumable_file_manager() -> "io.BufferedWriter":
...
+def _create_relative_symlink(src: str, dst: str) -> None:
+    """Create a symbolic link named dst pointing to src as a relative path to dst.
+
+    The relative part is mostly because it seems more elegant to the author.
+
+    The result layout looks something like
+        └── [ 128]  snapshots
+            ├── [ 128]  2439f60ef33a0d46d85da5001d52aeda5b00ce9f
+            │   ├── [  52]  README.md -> ../../blobs/d7edf6bd2a681fb0175f7735299831ee1b22b812
+            │   └── [  76]  pytorch_model.bin -> ../../blobs/403450e234d65943a7dcf7e05a771ce3c92faa84dd07db4ac20f592037a1e4bd
```
Comment (LysandreJik) [inline src/huggingface_hub/file_download.py:736]: Showcasing this is elegant, and I think showcasing the full layout somewhere is warranted as it makes understanding it super simple. Happy to provide it!

**Review by LysandreJik — COMMENTED** [review]
Thanks for working on it, this is super nice and clean.

It downloads files under `~/.cache/huggingface/hub`. Is this the expected place for them to be downloaded at? Should we expect other libraries (transformers, datasets) to go tap in that folder instead of the current transformers/datasets folders?

PS: When playing around with it, I had several connection errors that I didn't seem to have previously when using `snapshot_download`. Might be an isolated incident as I don't think this should put any more stress on the backend (quite the contrary).

File: src/huggingface_hub/file_download.py (hunk 12 above)
Comment (LysandreJik) [reply src/huggingface_hub/file_download.py:150]: (right now putting 4 characters means that we're skipping the cache check and going straight for `hf_hub_url`)

Comment (julien-c) [reply src/huggingface_hub/file_download.py:150]: oh i actually meant to only allow passing a full commit (i.e. just `{40}`), to make sure no false positive are detected (like naming a branch `deadbeef` etc)

Do you see an issue in accepting only full commit hashes?

File: src/huggingface_hub/snapshot_download.py [hunk 14]
```diff
@@ -120,133 +141,69 @@ def snapshot_download(
...
+    # we pass the commit_hash to hf_hub_download
+    # so no network call happens if we already
+    # have the file locally.
+
+    for repo_file in filtered_repo_files:
+        _ = hf_hub_download(
+            repo_id,
+            filename=repo_file,
+            repo_type=repo_type,
+            revision=commit_hash,
```
[the remark this answers is not in the document]
Comment (julien-c) [reply src/huggingface_hub/snapshot_download.py:198]: Yes i think i forgot to store the `ref` in `snapshot_download`, I realized that later. Will add it

**Comment by julien-c** [comment]
> It downloads files under `~/.cache/huggingface/hub`. Is this the expected place for them to be downloaded at? Should we expect other libraries (transformers, datasets) to go tap in that folder instead of the current transformers/datasets folders?

Yes, that was my idea – or we could also just go up one level and store in `~/.cache/huggingface`, considering that this is going to be library-agnostic and huggingface-wide anyways

**Comment by julien-c** [comment]
> PS: When playing around with it, I had several connection errors that I didn't seem to have previously when using `snapshot_download`. Might be an isolated incident

The API calls themselves are the same as before, so it shouldn't change much – let's investigate if it continues

File: src/huggingface_hub/snapshot_download.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (julien-c) [reply src/huggingface_hub/snapshot_download.py:198]: and actually your code suggestion is perfect i think

File: src/huggingface_hub/snapshot_download.py (hunk 6 above)
Comment (julien-c) [reply src/huggingface_hub/snapshot_download.py:182]: what do you mean?

 it should support arbitrary nesting, but i haven't extensively tested yet :)

File: src/huggingface_hub/file_download.py (hunk 9 above)
Comment (julien-c) [reply src/huggingface_hub/file_download.py:962]: yes 👍 on adding them

No strong opinion on whether we want to add them here or in another PR. I'd say another PR given we might want to also update `cached_download`, which this PR does not touch. WDYT @LysandreJik?

File: src/huggingface_hub/snapshot_download.py (hunk 3 above)
Comment (julien-c) [reply src/huggingface_hub/snapshot_download.py:158]: no strong opinion either way! i'll let @LysandreJik decide =)

File: src/huggingface_hub/file_download.py (hunk 13 above)
Comment (julien-c) [reply src/huggingface_hub/file_download.py:736]: took a stab at it in 81b75678b3556544f6a493c110057c8746a41c78

Did that using `tree -h`

File: src/huggingface_hub/file_download.py (hunk 10 above)
Comment (julien-c) [reply src/huggingface_hub/file_download.py:1131]: i'm not touching that particular line of code (just "unwrapping" it) so I'd rather keep it w/ same behavior as before

Comment (julien-c) [reply src/huggingface_hub/file_download.py:1131]: (to minimize scope of changes)

File: src/huggingface_hub/file_download.py [hunk 15]
```diff
@@ -771,21 +871,263 @@ def hf_hub_download(
...
+    # Prevent parallel downloads of the same file with a lock.
+    lock_path = blob_path + ".lock"
+
+    # Some Windows versions do not allow for paths longer than 255 characters.
+    # In this case, we must specify it is an extended path by using the "\\?\" prefix.
+    if os.name == "nt" and len(os.path.abspath(lock_path)) > 255:
+        lock_path = "\\\\?\\" + os.path.abspath(lock_path)
+
+    if os.name == "nt" and len(os.path.abspath(blob_path)) > 255:
+        blob_path = "\\\\?\\" + os.path.abspath(blob_path)
```
Comment (julien-c) [inline src/huggingface_hub/file_download.py:1081]: I'm not entirely certain this is still useful, TBH.... In any case, we're not doing it for other files or symlinks in this PR, so it's unlikely to still work as is

I'd consider removing it altogether @LysandreJik 

Slightly related poll on Windows older versions support: https://twitter.com/julien_c/status/1522958087883309059

File: src/huggingface_hub/snapshot_download.py (hunk 3 above)
Comment (LysandreJik) [reply src/huggingface_hub/snapshot_download.py:161]: Sure, I like this proposal! I find it improves readability, thank you @patrickvonplaten.

File: src/huggingface_hub/file_download.py (hunk 8 above)
Comment (LysandreJik) [reply src/huggingface_hub/file_download.py]: Handled in https://github.com/huggingface/huggingface_hub/pull/801/commits/d786693779f587c903b53adffe2b81734594edfd

File: src/huggingface_hub/file_download.py (hunk 9 above)
Comment (LysandreJik) [reply src/huggingface_hub/file_download.py:991]: Taking care of that in https://github.com/huggingface/huggingface_hub/pull/878

File: src/huggingface_hub/file_download.py (hunk 15 above)
Comment (LysandreJik) [reply src/huggingface_hub/file_download.py:1077]: This was setup for very recent windows versions as well. Will test on Windows to see if it can be removed.

It was introduced in https://github.com/huggingface/huggingface_hub/pull/40 due to https://github.com/huggingface/huggingface_hub/issues/35

**Comment by LysandreJik** [comment]
Remaining steps before merging:

- [ ] Write tests for the new cache layout
- [ ] Verify it works correctly on Windows
- [ ] Ensure backwards compatibility
- [ ] Document the layout

File: src/huggingface_hub/snapshot_download.py [hunk 16]
```diff
@@ -120,133 +141,90 @@ def snapshot_download(
...
+    # appropriate folder in the cache
+    # If the specified revision is a commit hash, look inside "snapshots".
+    # If the specified revision is a branch or tag, look inside "refs".
+    if local_files_only:
+        commit_hash = revision
+        if not REGEX_COMMIT_HASH.match(commit_hash):
+            # rertieve commit_hash from file
+            ref_path = os.path.join(storage_folder, "refs", revision)
+            with open(ref_path) as f:
+                commit_hash = f.read()
```
Comment (julien-c) [inline src/huggingface_hub/snapshot_download.py:166]:
```suggestion
        def resolve_ref(revision) -> str:
            # retrieve commit_hash from file
            ref_path = os.path.join(storage_folder, "refs", revision)
            with open(ref_path) as f:
                return f.read()

        commit_hash = revision 
            if REGEX_COMMIT_HASH.match(revision)
            else resolve_ref(revision)
```

I think that maybe it is cleaner to use a ternary then @LysandreJik @patrickvonplaten WDYT?

**Comment by thomwolf** [comment]
I noticed a bug here when using `subfolder`:

If I have two files with the same name in different subfolders and I'm downloading them subsequently using the `hf_hub_download` method, the second file will not be downloaded because it's seen as already existing.

A space to test the issue: these two files are different (see the end of the gltf files):
- https://huggingface.co/spaces/simenv-tests/Box/blob/main/glTF-Embedded/Box.gltf
- https://huggingface.co/spaces/simenv-tests/Box/blob/main/glTF/Box.gltf

Proposal to solve the issue: add the `subfolder` in the path used to create the cache file in the snapshot folder so that `if os.path.exists(pointer_path)` fails instead or giving true.

Current workaround I'll use: set `force_download` to True

**Comment by julien-c** [comment]
can you post the snippet of code you use?

**Comment by julien-c** [comment]
I think there's a bug in this line:
```python
# cross platform transcription of filename, to be used as a local file path.
relative_filename = os.path.join(*filename.split("/"))
```
where it should take into account the potential `subfolder` param

**Comment by thomwolf** [comment]
Code snippet could be for instance:
```
first_file = hf_hub_download(repo_id='simenv-tests/Box',
            filename='Box.gltf,
            subfolder='glTF',
            repo_type="space",
        )
print(first_file)
second_file = hf_hub_download(repo_id='simenv-tests/Box',
            filename='Box.gltf,
            subfolder='glTF-Embedded',
            repo_type="space",
        )
print(second_file)
```
Manual inspection of the first and second files show they are identical (see in particular the `"buffers"` section of both JSON)

If you remove the `subfolder` parameter, what would you replace it with?

**Comment by julien-c** [comment]
> If you remove the `subfolder` parameter, what would you replace it with?

just the filename param, should be the full nested filename: `glTF/Box.gltf `

File: tests/test_cache_layout.py [hunk 17]
```diff
@@ -0,0 +1,385 @@
...
+        repo_id = f"{USER}/hfh-cache-layout"
+        create_repo(repo_id, token=self._token, exist_ok=True)
+
+        try:
+            upload_file(
+                path_or_fileobj=BytesIO(b"Some string"),
+                path_in_repo="file.txt",
+                repo_id=repo_id,
+                token=self._token,
+            )
```
Comment (julien-c) [inline tests/test_cache_layout.py:324]: this method is soon to be deprecated but we can update it then (cc @SBrandeis)

File: tests/test_cache_layout.py [hunk 18]
```diff
@@ -0,0 +1,385 @@
...
+    def setUpClass(cls):
+        """
+        Share this valid token in all tests below.
+        """
+        cls._token = TOKEN
+        cls._api.set_access_token(TOKEN)
+
+    def test_update_reference(self):
+        repo_id = f"{USER}/hfh-cache-layout"
+        create_repo(repo_id, token=self._token, exist_ok=True)
```
Comment (julien-c) [inline tests/test_cache_layout.py:316]: not sure about this test: it's not going to be able to run it multiple times concurrently

Is the plan to have a uuid in the repo name like in other "integration tests"?

File: tests/test_cache_layout.py [hunk 19]
```diff
@@ -0,0 +1,385 @@
...
+            # The snapshot content should link to a blob
+            self.assertTrue(os.path.islink(snapshot_content_path))
+
+            resolved_blob_relative = os.readlink(snapshot_content_path)
+            resolved_blob_absolute = os.path.normpath(
+                os.path.join(snapshot_path, resolved_blob_relative)
+            )
+
+            with open(resolved_blob_absolute) as f:
+                blob_contents = f.readline().strip()
```
[the remark this answers is not in the document]
Comment (LysandreJik) [reply tests/test_cache_layout.py:72]: Because the file contains a single line! Reading the whole file would be better, indeed. Thanks!

File: tests/test_cache_layout.py (hunk 18 above)
Comment (LysandreJik) [reply tests/test_cache_layout.py:316]: Yes, you're correct. Updating to use the uuid.

File: src/huggingface_hub/file_download.py [hunk 20]
```diff
@@ -514,6 +526,11 @@ def cached_download(
 
     </Tip>
     """
+    warnings.warn(
+        "`cached_download` is the legacy way to download files from the HF hub, please"
+        " consider upgrading to `hf_hub_download`",
+        FutureWarning,
+    )
```
Comment (LysandreJik) [inline src/huggingface_hub/file_download.py:533]: @julien-c, `cached_download` implements features that are not available with `hf_hub_download` (namely, caching from a URL). Is this a feature that you think should be abandonned? I think there is room for both methods to exist, but this warning says otherwise.

Comment (julien-c) [reply src/huggingface_hub/file_download.py:533]: > Is this a feature that you think should be abandonned?

I think it's a feature that does not really make sense in the context of the HF hub, and a HF Hub client library.

My concern is that if we don't add this warning, users will not upgrade to the new way.

As a middleground, maybe we could add a `legacy_cache_layout` flag that just disables the warning? and mention in the warning "If you want to cache from an arbitrary URL instead, pass legacy_cache_layout=True"

WDYT?

File: src/huggingface_hub/_snapshot_download.py [hunk 21]
```diff
@@ -120,133 +141,90 @@ def snapshot_download(
...
+    # appropriate folder in the cache
+    # If the specified revision is a commit hash, look inside "snapshots".
+    # If the specified revision is a branch or tag, look inside "refs".
+    if local_files_only:
+        commit_hash = revision
+        if not REGEX_COMMIT_HASH.match(commit_hash):
+            # rertieve commit_hash from file
+            ref_path = os.path.join(storage_folder, "refs", revision)
+            with open(ref_path) as f:
+                commit_hash = f.read()
```
Comment (julien-c) [inline src/huggingface_hub/_snapshot_download.py:166]: still need to update this one no?

File: src/huggingface_hub/file_download.py (hunk 20 above)
Comment (LysandreJik) [reply src/huggingface_hub/file_download.py:533]: Yes, sounds good to me. I'll proceed this way.

File: src/huggingface_hub/_snapshot_download.py [hunk 22]
```diff
@@ -158,26 +158,20 @@ def snapshot_download(
...
             ref_path = os.path.join(storage_folder, "refs", revision)
             with open(ref_path) as f:
-                commit_hash = f.read()
+                return f.read()
 
+        commit_hash = (
+            revision if REGEX_COMMIT_HASH.match(revision) else resolve_ref(revision)
+        )
         snapshot_folder = os.path.join(storage_folder, "snapshots", commit_hash)
+
```
Comment (LysandreJik) [inline src/huggingface_hub/_snapshot_download.py:172]: See side by side comparison of before/after @julien-c:


<table>
<tr>
<td>After</td>
<td>Before</td>
</tr>

<tr>
<td>

```py
if local_files_only:

    def resolve_ref(revision) -> str:
        # retrieve commit_hash from file
        ref_path = os.path.join(storage_folder, "refs", revision)
        with open(ref_path) as f:
            return f.read()

    commit_hash = (
        revision if REGEX_COMMIT_HASH.match(revision) else resolve_ref(revision)
    )
    snapshot_folder = os.path.join(storage_folder, "snapshots", commit_hash)

    if os.path.exists(snapshot_folder):
        return snapshot_folder

    raise ValueError(
        "Cannot find an appropriate cached snapshot folder for the specified"
        " revision on the local disk and outgoing traffic has been disabled. To"
        " enable repo look-ups and downloads online, set 'local_files_only' to"
        " False."
    )
```

</td>
<td>

```py
if local_files_only:
    if REGEX_COMMIT_HASH.match(revision):
        snapshot_folder = os.path.join(storage_folder, "snapshots", revision)
        if os.path.exists(snapshot_folder):
            return snapshot_folder
    else:
        ref_path = os.path.join(storage_folder, "refs", revision)
        with open(ref_path) as f:
            commit_hash = f.read()
        snapshot_folder = os.path.join(storage_folder, "snapshots", commit_hash)
        if os.path.exists(snapshot_folder):
            return snapshot_folder
     raise ValueError(
        "Cannot find an appropriate cached folder for the specified revision on the"
        " local disk and outgoing traffic has been disabled. To enable repo"
        " look-ups and downloads online, set 'local_files_only' to False."
    )
```

</td>
</tr>

</table>

Comment (julien-c) [reply src/huggingface_hub/_snapshot_download.py:172]: Yep, looks great to me!

File: docs/source/package_reference/file_download.mdx [hunk 23]
```diff
@@ -8,3 +8,84 @@
...
+
+The caching system is designed as follows:
+
+```
+<CACHE_DIR>
+├─ <MODELS>
+├─ <DATASETS>
+├─ <SPACES>
+```
+
```
Comment (sgugger) [inline docs/source/package_reference/file_download.mdx:25]: Maybe remind the user the default for `CACHE_DIR` and how it can be customized?

File: docs/source/package_reference/file_download.mdx [hunk 24]
```diff
@@ -8,3 +8,84 @@
...
+Each folder is designed to contain the following:
+
+### Refs
+
+The `refs` folder contains files which indicates the latest revision of the given reference. For example,
+if we have previously fetched a file from the `main` branch of a repository, the `refs`
+folder will contain a file named `main`, which will itself contain the commit identifier of the current head.
+
+If the latest commit of `main` had `aaaaaa` as identifier, then it would contain `aaaaaa`.
```
Comment (sgugger) [inline docs/source/package_reference/file_download.mdx:63]:
```suggestion
If the latest commit of `main` has `aaaaaa` as identifier, then it will contain `aaaaaa`.
```

File: docs/source/package_reference/file_download.mdx [hunk 25]
```diff
@@ -8,3 +8,84 @@
...
+if we had downloaded the `READMD.md` file at revision `aaaaaa`, we would have the following path:
+
+```
+<CACHE_DIR>/<REPO_NAME>/snapshots/aaaaaa/README.md
+```
+
+That `README.md` file is actually a symlink linking to the blob that has the hash of the file.
+
+Creating the skeleton this way means opens up the mechanism to file sharing: if the same file was fetched in
+revision `bbbbbb`, it would have the same hash and the file would not need to be redownloaded.
```
Comment (sgugger) [inline docs/source/package_reference/file_download.mdx:91]: I think it would be helpful to include the example @julien-c put in the description to illustrate all the theoretical doc.

File: src/huggingface_hub/_snapshot_download.py [hunk 26]
```diff
@@ -120,133 +141,84 @@ def snapshot_download(
...
-        # if there's an allowlist, skip download if file does not match any regex
-        if allow_regex is not None and not any(
-            fnmatch(model_file, r) for r in allow_regex
-        ):
-            continue
+        def resolve_ref(revision) -> str:
+            # retrieve commit_hash from file
+            ref_path = os.path.join(storage_folder, "refs", revision)
+            with open(ref_path) as f:
+                return f.read()
```
Comment (sgugger) [inline src/huggingface_hub/_snapshot_download.py:166]: This function is used once so I don't get why we refactor the three lines of code in it. E.g.,
```py
if REGEX_COMMIT_HASH.match(revision):
    commit_hash = revision
else:
    # retrieve commit_hash from file
    ref_path = os.path.join(storage_folder, "refs", revision)
    with open(ref_path) as f:
        commit_hash = f.read()
```

File: src/huggingface_hub/file_download.py [hunk 27]
```diff
@@ -689,6 +730,73 @@ def _resumable_file_manager() -> "io.BufferedWriter":
...
+            )
+        else:
+            raise
+
+
+def repo_folder_name(
+    *,
+    repo_id: str,
+    repo_type: str,
+) -> str:
```
Comment (sgugger) [inline src/huggingface_hub/file_download.py:786]: Think even with a 80 char limit this fits in one line ;-)

File: src/huggingface_hub/file_download.py [hunk 28]
```diff
@@ -689,6 +730,73 @@ def _resumable_file_manager() -> "io.BufferedWriter":
...
+    """Return a serialized version of a hf.co repo name and type, safe for disk storage
+    as a single non-nested folder.
+
+    Example: models--julien-c--EsperBERTo-small
+    """
+    # remove all `/` occurrences to correctly convert repo to directory name
+    parts = [
+        f"{repo_type}s",
+        *repo_id.split("/"),
+    ]
```
Comment (sgugger) [inline src/huggingface_hub/file_download.py:796]: Same here.

File: src/huggingface_hub/file_download.py [hunk 29]
```diff
@@ -753,6 +886,10 @@ def hf_hub_download(
         local_files_only (`bool`, *optional*, defaults to `False`):
             If `True`, avoid downloading the file and return the path to the
             local cached file if it exists.
+        legacy_cache_layout (`bool`, *optional*, defaults to `False`):
+            If `True`, uses the legacy file cache layout i.e. just call `hf_hub_url`
```
Comment (sgugger) [inline src/huggingface_hub/file_download.py:890]:
```suggestion
            If `True`, uses the legacy file cache layout i.e. just call [`hf_hub_url`]
```
Link to `hf_hub_url`?

File: src/huggingface_hub/file_download.py [hunk 30]
```diff
@@ -771,21 +908,269 @@ def hf_hub_download(
...
+
+    if cache_dir is None:
+        cache_dir = HUGGINGFACE_HUB_CACHE
+    if revision is None:
+        revision = DEFAULT_REVISION
+    if isinstance(cache_dir, Path):
+        cache_dir = str(cache_dir)
+
+    if subfolder is not None:
+        filename = f"{subfolder}/{filename}"
```
Comment (sgugger) [inline src/huggingface_hub/file_download.py:952]: Maybe put a comment here that we use `/` since it is used for URLs or local paths, and that we will deal with / on Windows afterward?
(I did a comment on Windows that I deleted after seeing line 967 below.)

File: src/huggingface_hub/file_download.py [hunk 31]
```diff
@@ -771,21 +908,269 @@ def hf_hub_download(
...
+            r.raise_for_status()
+            commit_hash = r.headers[HUGGINGFACE_HEADER_X_REPO_COMMIT]
+            if commit_hash is None:
+                raise OSError(
+                    "Distant resource does not seem to be on huggingface.co (missing"
+                    " commit header)."
+                )
+            etag = r.headers.get(HUGGINGFACE_HEADER_X_LINKED_ETAG) or r.headers.get(
+                "ETag"
+            )
```
Comment (sgugger) [inline src/huggingface_hub/file_download.py:1021]:
```suggestion
            etag = r.headers.get(HUGGINGFACE_HEADER_X_LINKED_ETAG)
            etag = etag or r.headers.get("ETag")
```
Probably a matter of personal preference, but I find it easier to read like this.

**Review by sgugger — APPROVED** [review]
Left some nits, feel free to ignore.
Very nice tests!

File: src/huggingface_hub/file_download.py (hunk 31 above)
Comment (julien-c) [reply src/huggingface_hub/file_download.py:1021]: you get the huge performance hit of doing two var assignments instead of one, and it's a bit less "functional" (well, it's Python, so...) 😅

(no strong opinion though)

**Comment by LysandreJik** [comment]
Thank you, @patrickvonplaten, @sgugger, @julien-c and @thomwolf for the reviews! I'll go ahead and merge this now so that we can test it extensively while on `main` to ensure nothing breaks.
